plugins/logitech-bulkcontroller/fu-plugin-logitech-bulkcontroller.c

```diff
@@ -0,0 +1,291 @@
+
+        user_data.prog_name = g_strdup ("Logitech Rally Bar Mini");
+        user_data.device_ptr = (FuDevice *)device;
+        obj = logibulkcontroller_create_bulk_controller (0x46d, 0x8d3, bulkcb, &user_data);
+        if (ERRORCODE_NO_ERROR != logibulkcontroller_open_device (obj)) {
+                g_warning("Unknown device\n");
+                return 0;
+        }
+
+        gchar * version_string = get_device_version(obj, &user_data);
+        g_print ("\n SRS version_string 1st time %s: \n", version_string);
+
+        g_print ("\n SRS Calling logibulkcontroller_send_file_upd \n");
+        fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+        logibulkcontroller_send_file_upd (obj, blob_fw, g_bytes_get_size(blob_fw), FALSE);
+        g_print ("\n SRS Called logibulkcontroller_send_file_upd \n");
+        if (g_cond_wait_until (&user_data.test_upd_cond, &user_data.test_upd_mutex,
+                                                        g_get_monotonic_time () + TIMEOUT_20M * G_TIME_SPAN_SECOND)) {
+                g_mutex_unlock (&user_data.test_upd_mutex);
+                g_mutex_lock (&user_data.test_upd_mutex);
+        }
+        fu_device_set_status (device, FWUPD_STATUS_DEVICE_VERIFY);
+        g_print ("\n SRS Calling logibulkcontroller_send_data_sync again \n");
+        version_string = get_device_version(obj, &user_data);
+                g_print ("\n SRS version_string 2nd time %s: \n", version_string);
+        sleep(3); //For testing
+        
+        g_print ("\n SRS Done with g_cond_wait_until \n");
+        logibulkcontroller_close_device (obj);
+        g_print ("\n SRS Done with logibulkcontroller_close_device \n");
+
+        fu_device_set_version_format (device, FWUPD_VERSION_FORMAT_TRIPLET);
+        fu_device_set_version (device, "1.2.3");
+
+	g_print ("\n SRS fu_plugin_update Exit \n");
+        
+        return TRUE;
+}
+
+void
+fu_plugin_init (FuPlugin *plugin)
+{
+	fu_plugin_set_build_hash (plugin, FU_BUILD_HASH);
+	g_debug ("Initializing Logitech bulk controller plugin\n");
+}
+
+void
+fu_plugin_destroy (FuPlugin *plugin)
+{
+        g_debug ("Terminating Logitech bulk controller plugin\n");
+}
+
+gboolean
+fu_plugin_coldplug (FuPlugin *plugin, GError **error)
+{
+        g_print ("\n SRS fu_plugin_coldplug entry \n");
+	FuContext *ctx = fu_plugin_get_context (plugin);
+	g_autoptr(FuDevice) device = NULL;
+	device = fu_device_new_with_context (ctx);
+	fu_device_set_id (device, "FakeDevice");
+	fu_device_add_guid (device, "b585990a-003e-5270-89d5-3705a17f9a43");
+	fu_device_set_name (device, "Rally Bar Mini");
+	//fu_device_add_icon (device, "preferences-desktop-keyboard");
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_protocol (device, "com.acme.test");
+	fu_device_set_vendor (device, "Logitech");
+	fu_device_add_vendor_id (device, "USB:0x046D");
+	fu_device_set_version_format (device, FWUPD_VERSION_FORMAT_TRIPLET);
+	fu_device_set_version (device, "1.2.3");
```

Comment:
remove the hardcoded version

Reply:
I need some help here understanding how devices are discovered. I am leveraging test plugin code and tutorial document.
We need to support 2 devices, Rally Bar (USB\VID_046D&PID_089B) and Rally Bar Mini (USB\VID_046D&PID_08D3). Firmware files are different for both files. We need to understand which of this device is connected currently so that we can use appropriate PID to communicate with that device using gusb. How does our plugin determine which devices are connected? Do we probe each of our supported devices to see if they are connected? We do this in fu_plugin_coldplug() vfunc?

In test plugin, when "fu_plugin_test_get_version (GBytes *blob_fw)" callback gets invoked? What is blob_fw arg here?
We have implemented our API to get currently installed firmware version from the device. Which callback (vfunc) we have to implement to call our get version API